### PR TITLE
camera_request

### PR DESCRIPTION
**Pull request by grecubogdan-13**

### CODE REVIEW

**Comment by grecubogdan-13** [comment]
mai am putin de lucru

**Comment by grecubogdan-13** [comment]
pusesem extensia proasta .TS nu .ts :)

File: src/requests.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import {Param} from "./Interfaces";
+const api_key = "TxbNgclp1JXLSnghP5ecigSdEUvdMNcabY3p1lno";
+const base_api = "https://api.nasa.gov/mars-photos/api/v1/rovers/";
```
Comment (malinaslevoaca1) [inline src/requests.ts]: this is not the base API url but the rovers one

also, in general the constants are all caps

File: src/server.ts [hunk 2]
```diff
@@ -14,105 +17,81 @@ app.listen(port, () => {
...
         console.log("agrethsh");
     });
 });
 
-app.get('/rovers', (req, res)=>{
-    let api_url = `https://api.nasa.gov/mars-photos/api/v1/rovers?api_key=${api_key}`
+app.get('/rovers/:rover_name/sol/:sol/photos/html', (req, res)=> {
+    let parameters: Param = {};
+    parameters.name = req.params.rover_name;
+    parameters.sol = req.params.sol;
```
Comment (malinaslevoaca1) [inline src/server.ts]: you could have 
```
parameters = {
    name: req.params.rover_name,
    sol: req.params.sol
}

File: src/JSONResponseFunctions.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import {
+    JSONInterfacePhotosBySolRequest,
+    JSONInterfacePhotosBySolResponse, RoversCamera,
+    RoversInterfaceRequest,
+    RoversInterfaceResponse
+} from "./Interfaces";
+
+
+export function photosJSON(photoArray: JSONInterfacePhotosBySolRequest[]) {
+    let response = '{ "photos":[';
```
Comment (malinaslevoaca1) [inline src/JSONResponseFunctions.ts]: why is this?

File: src/JSONResponseFunctions.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            earth_date: string = "";
+            img_src: string = "";
+            photo_id: number = 0;
+            rover_name: string = "";
+        };
+        selectData.photo_id = d.id;
+        selectData.rover_name = d.rover.name;
+        selectData.img_src = d.img_src;
+        selectData.earth_date = d.earth_date;
+        selectData.camera_name = d.camera.full_name;
```
Comment (malinaslevoaca1) [inline src/JSONResponseFunctions.ts]: is there a reason why you haven't added these above?

File: package.json [hunk 5]
```diff
@@ -15,7 +15,7 @@
   "devDependencies": {},
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "start": "nodemon src/server.ts"
+    "start": "nodemon src/Routes/routes.ts"
```
Comment (malinaslevoaca1) [inline package.json:18]: we shouldn't make this change.
the server configuration should live in `server.ts`, only the routes to be moved to a separate file

File: src/Controllers/controller.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
+import {requestConstructor} from "../Services/Requests/requests";
+import axios from "axios";
+import {photosJSON, roversJSON} from "../Services/Responses/JSONResponseFunctions";
+import {photosHTML, roversHTML} from "../Services/Responses/HTMLResponseFunctions";
+import {Request, Response} from "express";
+
+export const test = (req: Request, res: Response) => {
```
Comment (malinaslevoaca1) [inline src/Controllers/controller.ts:7]: in general we have a controller for each resource type - test will live in `testController`
all the endpoints related to rovers in `roversController` etc

File: src/Controllers/controller.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import axios from "axios";
+import {photosJSON, roversJSON} from "../Services/Responses/JSONResponseFunctions";
+import {photosHTML, roversHTML} from "../Services/Responses/HTMLResponseFunctions";
+import {Request, Response} from "express";
+
+export const test = (req: Request, res: Response) => {
+    res.send('Hello world !');
+}
+
+export const JSONNameAndSol = async (req: Request, res: Response) => {
```
Comment (malinaslevoaca1) [inline src/Controllers/controller.ts:11]: the name could be more descriptive

File: src/Controllers/controller.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    let parameters = {
```
Comment (malinaslevoaca1) [inline src/Controllers/controller.ts:12]: usually if we don't reassign a variable we declare it as const

File: src/Controllers/controller.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        name: req.params.rover_name,
+        sol: req.params.sol
+    }
+    let api_url = requestConstructor(parameters, "RoverAndSol");
+    if (api_url) {
+        let resp = await axios.get(api_url);
+        let data = resp.data;
+        let response = photosJSON(data.photos)
```
Comment (malinaslevoaca1) [inline src/Controllers/controller.ts:20]: could move this logic to a nasaApiService (service used for fetching data)

File: src/Models/Interfaces.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
+export interface JSONInterfacePhotosBySolRequest {
```
Comment (malinaslevoaca1) [inline src/Models/Interfaces.ts:1]: usually each interface lives in its own file

you don't need `JsonInterface` in the interface name, do you?

File: src/Services/Responses/JSONResponseFunctions.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    JSONInterfacePhotosBySolResponse, JSONInterfacePhotosList, RoversCamera,
+    RoversInterfaceRequest,
+    RoversInterfaceResponse
+} from "../../Models/Interfaces";
+
+export function photosJSON(photoArray: JSONInterfacePhotosBySolRequest[]) {
+    let response = new class implements JSONInterfacePhotosList {
+        photos: Array<JSONInterfacePhotosBySolResponse> = [];
+    };
+    photoArray.forEach(function (d: JSONInterfacePhotosBySolRequest) {
```
Comment (malinaslevoaca1) [inline src/Services/Responses/JSONResponseFunctions.ts:12]: could use `map` instead

**Review by malinaslevoaca1 — APPROVED** [review]
Some suggestions below - no need to implement them unless you run out of work

**Pull request merged**